### PR TITLE
Added fix for "same" padding in single output mode. Importnat for compatibility with TFLite.

### PR DESCRIPTION
**Pull request by karolbadowski**
Added fix enabling full deductive capabilities with usage of 'same' padiing in "one output" mode. 
In case of 'same' padding, slice we want to return is one at middle index of last layer. 
Otherwise neural network will not reach results comparable in quality to 'causal' padding. 

Why 'same' padding is important? 
Currently it is the only one padding in dilated convolution that is supported by TensrFlow-Lite #TFLite.

### CODE REVIEW

**Comment by philipperemy** [comment]
@karolbadowski I can't really test it with TFLite. How confident are you with this PR? :)

**Comment by karolbadowski** [comment]
TFLite uses converter named TOCO in order to convert from Keras model to TFLite model.
Currently it does not support dilated convolutions with "causal" padding ( https://github.com/tensorflow/tensorflow/pull/28410 ). It just ends with conversion error due to lack of support for such scenario in current TFLite version.

In order to emulate network as similar to basic TCN ("causal", singe output), you need to express similar shape of connections with usage of padding "same".

Theoretically, when you:
- do not use skip connections,
- and you have 1 stack of residual blocks
- you organise input the same way as in classical TCN
you achieve a neural network of analogical shape and analogical deductive capabilities (it should be able to train equally well).

In practice, I have not been able to achieve results close in quality.
Upon analysis of code of this library, I have noticed the reason.

For "same" padding, everything is just OK with what is returned upon "multiple outputs" mode.
However in "single output" mode, the returned index should be in the middle.  Otherwise neural network can see only half of the receptive field.

**Comment by karolbadowski** [comment]
I like your analogy that "same" padding is good when you consider problems of analyzing moments of the past and future. It is prominent when it comes to alignment of places where skip connections are joined with new layers. In your idea, the organisation of data is different, because last half of receptive field show moments of the future.

Nevertheless, even in this scenario, when you want to return only 1 output equivalent to current moment, you should output the one which has ID: [LAST_ID - half of receptive field size]. **In other words it is the last ID that has seen entire receptive field.**

**Comment by karolbadowski** [comment]
I would like to revive the topic.

In case of "same" padding, when we return single output instead of a sequence,
that output is in the middle of last layer (not on last position as it would be in "causal" padding).

**Comment by philipperemy** [comment]
@karolbadowski Thanks I need to prioritize this one :)

**Comment by karolbadowski** [comment]
I see there were changes that produce a conflict.
I believe that to be compliant with new changes, now my fix would need to be applied in build() method instead of call() method.

Conflict resolving by changing the ending of build() method in following way:
```
        self.lambda_ouput_shape = self.lambda_layer.compute_output_shape(self.build_output_shape)
        output_slice_index = int(self.lambda_ouput_shape.as_list()[1]/2) if self.padding == 'same' else -1
        self.lambda_layer = Lambda(lambda tt: tt[:, output_slice_index, :])
```

Please confirm if with newest changes self.lambda_ouput_shape is a tensor or a tuple (so that we know if .as_list() is still necessary).

**Comment by philipperemy** [comment]
@karolbadowski second one is better in my opinion, as you suggested. I made a PR to reflect your change: https://github.com/philipperemy/keras-tcn/pull/103. I added you as author.

**Comment by philipperemy** [comment]
Thanks again by the way! Sorry I was busy those past few months with TF2.0 support :)

**Comment by philipperemy** [comment]
@karolbadowski `self.build_output_shape` should be a `TensorShape`

**Comment by philipperemy** [comment]
@karolbadowski so I made a PR here: https://github.com/philipperemy/keras-tcn/pull/103.
You can either update your PR with the changes here or give me the green light to merge it.
Thank you!

**Comment by philipperemy** [comment]
Ok going to merge the other one :)

**Comment by karolbadowski** [comment]
@philipperemy sorry for delay in response, that was a busy week. 
Thank you for the update :).
